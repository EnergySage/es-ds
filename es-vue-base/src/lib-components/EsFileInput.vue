<template>
    <div
        role="presentation"
        class="es-file-upload align-items-center rounded d-flex flex-column justify-content-center p-2"
        :class="{ 'active': active }"
        @dragenter.stop.prevent="active = true"
        @dragover.stop.prevent="active = true"
        @dragleave.stop.prevent="active = false"
        @dragend.stop.prevent="onDrop"
        @drop.stop.prevent="onDrop"
        @click.self="openFilePicker"
        @keypress.enter.prevent="openFilePicker"
        @keypress.space.prevent="openFilePicker">
        <div
            class="d-flex justify-content-center flex-column"
            :class="{ 'flex-md-row': collapsed }">
            <div class="align-self-center p-2">
                <icon-upload
                    class="d-none d-md-inline-block mb-4"
                    height="48px"
                    width="48px" />
            </div>
            <div class="align-self-center p-2 text-center">
                <slot name="cta" />
            </div>
            <div class="align-self-center p-2">
                <es-button
                    class="w-100 w-md-auto mb-2 mb-md-4"
                    variant="secondary"
                    outline
                    @click="openFilePicker">
                    <slot name="buttonText">
                        <p class="d-none d-md-inline m-0">
                            Browse files
                        </p>
                        <p class="d-md-none m-0">
                            Choose file
                        </p>
                    </slot>
                </es-button>
            </div>
        </div>
        <div>
            <slot name="helpText" />
        </div>
        <b-form-file
            ref="fileInput"
            v-model="pickedItems"
            :state="Boolean(pickedItems)"
            :accept="fileTypes.join(', ')"
            style="display: none"
            multiple />
    </div>
</template>

<script lang="js">
import { BFormFile } from 'bootstrap-vue';
import EsButton from './EsButton.vue';
import IconUpload from '../lib-icons/icon-upload.vue';
import { mimeTypes, findMimeType } from '../lib-utils';

export default {
    name: 'EsFileInput',
    components: {
        BFormFile,
        EsButton,
        IconUpload,
    },
    props: {
        /**
         * An array of Objects with the following shape:
         * {
         *    fileName: String,
         *    uploadUrl: String,
         *    additionalFields: Object,
         * }
         */
        uploadUrls: {
            type: Array,
            default: () => [],
        },
        /**
         * The file types that are allowed to be uploaded. This is a list of MIME types as strings. If the array is
         * empty, all file types are allowed. A list of acceptable MIME types can be found here in the template column:
         * https://www.iana.org/assignments/media-types/media-types.xhtml
         */
        fileTypes: {
            type: Array,
            default: () => [],
            required: true,
            validator(fileTypes) {
                return fileTypes.every((fileType) => mimeTypes.includes(fileType));
            },
        },
        /**
         * The maximum file size in MB. This is per file, not the total size of all files.
         */
        maxFileSizeMb: {
            type: Number,
            default: 25,
            required: false,
        },
        /**
         * When true, the component will be shorter vertically. This is achieved by giving the parent div
         * flex-direction: row instead of the default flex-direction: column.
         */
        collapsed: {
            type: Boolean,
            default: false,
            required: false,
        },
        deleteFileName: {
            type: String,
            default: '',
            required: false,
        },
    },
    data() {
        return {
            pickedItems: [],
            files: [],
            active: false,
        };
    },
    watch: {
        uploadUrls: {
            handler(newUrls, oldUrls) {
                // For every file that has an associated upload URL, we start the upload
                newUrls.forEach((newUrlPair) => {
                    const oldUrl = oldUrls.find(({ name }) => name === newUrlPair.name);
                    const fileToUpload = this.files.find((file) => file.name === newUrlPair.name);
                    if ((!oldUrl || oldUrl.uploadUrl !== newUrlPair.uploadUrl) && fileToUpload) {
                        this.uploadSingleFile(fileToUpload);
                    }
                });
            },
            deep: true,
        },
        async pickedItems(newVal) {
            // The user has selected files from the file picker. We have to filter out any files that exceed the
            // maxFileSizeMb prop.
            if (newVal.length > 0) {
                await this.verifyFiles(newVal);
            }
        },
        deleteFileName(newVal) {
            this.files = this.files.filter((file) => file.name !== newVal);
        },
    },
    methods: {
        removeSpaceFromFileNames(files) {
            const newFiles = files.map((file) => new File(
                // Return new File object with the modified name (without any space)
                [file],
                file.name.replace(/\s/g, ''),
                { type: file.type, lastModified: file.lastModified },
            ));
            return newFiles;
        },
        filterLargeFiles(files) {
            // Takes an array of files, and filters out any files that exceed the maxFileSizeMb prop. Emits a
            // fileSizeError event for each file that exceeds the maxFileSizeMb prop. The File API uses bytes, so
            // we have to convert the maxFileSizeMb prop to bytes.
            const maxFileSizeBytes = this.maxFileSizeMb * 1000000;
            return files.filter((file) => {
                if (file.size > maxFileSizeBytes) {
                    this.$emit('fileSizeError', file.name);
                    return false;
                }
                return true;
            });
        },
        readFilesIntoUrl(files) {
            files
                .forEach((file) => {
                    const fileReader = new FileReader();
                    fileReader.onload = () => {
                        this.$emit('fileDataRead', {
                            name: file.name,
                            type: file.type,
                            size: file.size,
                            data: file.type.includes('application')
                                // For non-image files (pdf, docx, etc.)
                                ? URL.createObjectURL(new Blob([file], { type: file.type }))
                                : fileReader.result,
                        });
                    };
                    fileReader.readAsDataURL(file);
                });
        },
        async verifyFiles(files) {
            const correctlySizedFiles = this.filterLargeFiles(files);
            const correctlyNamedFiles = this.removeSpaceFromFileNames(correctlySizedFiles);

            // Make sure the file is the correct mime type
            let newValidFiles = await Promise.allSettled(
                correctlyNamedFiles.map(async (file) => this.verifyMimeType(file)),
            );
            // Filter out undefined values and "rejected"
            newValidFiles = newValidFiles.filter((file) => file && file.status === 'fulfilled')
                .map((file) => file.value);

            // If the new file with name already exists in the current files
            // Don't upload the new file and display an error
            const duplicateFiles = newValidFiles.filter(({ name }) => this.files.some((file) => file.name === name));
            duplicateFiles.forEach((file) => this.$emit('duplicateFileMessage', file.name));

            // For new files with unused names, add them to files
            newValidFiles = newValidFiles.filter(({ name }) => !this.files.some((file) => file.name === name));
            this.files = [...this.files, ...newValidFiles];

            if (newValidFiles.length > 0) {
                this.$emit('readyToUpload', newValidFiles);
                this.readFilesIntoUrl(newValidFiles);
            }
            this.pickedItems = [];
        },
        async verifyMimeType(file) {
            // If an empty folder then trigger fileTypeError
            if (file.type === '') {
                this.$emit('fileIsAFolderError');
            }
            return new Promise((resolve, reject) => {
                const fileReader = new FileReader();
                fileReader.onload = (evt) => {
                    const uint = new Uint8Array(evt.target.result);
                    const bytes = [];
                    uint.forEach((byte) => {
                        bytes.push(byte.toString(16));
                    });
                    const hex = bytes.join('').toUpperCase();
                    const mimeType = findMimeType(hex);
                    if (!mimeType || (this.fileTypes.length > 0 && !this.fileTypes.includes(mimeType))) {
                        this.$emit('fileTypeError', file.name);
                        return reject();
                    }
                    return resolve(file);
                };
                fileReader.onerror = (error) => {
                    reject(error);
                };
                const blob = file.slice(0, 4);
                fileReader.readAsArrayBuffer(blob);
            });
        },
        onDrop(event) {
            // The user has dropped files onto the component. We have to apply the same logic as if they had
            // selected the files from the file picker which limits the file types to the ones specified in the
            // fileTypes prop.
            this.active = false;

            // Use DataTransferItemList interface to access the file(s)
            const dataTransfersAsFiles = [...event.dataTransfer.items]
                .filter((item) => {
                    if (item.kind !== 'file') {
                        this.$emit('fileTypeError', item.name);
                        return false;
                    }
                    return true;
                })
                .map((item) => item.getAsFile());

            this.verifyFiles(dataTransfersAsFiles);
        },
        openFilePicker() { this.$refs.fileInput.$el.childNodes[0].click(); },
        async uploadSingleFile(file) {
            const uploadInfo = this.uploadUrls.find((uploadUrl) => uploadUrl.name === file.name);
            if (!uploadInfo) {
                this.$emit('uploadFailure', {
                    name: file.name,
                    message: 'No upload URL found for this file.',
                });
                return;
            }
            const form = new FormData();
            if (uploadInfo.additionalFields) {
                Object.entries(uploadInfo.additionalFields).forEach(([key, value]) => {
                    form.append(key, value);
                });
            }
            form.append('file', file);
            fetch(uploadInfo.uploadUrl, {
                method: 'POST',
                body: form,
            })
                .then((response) => {
                    if (response.status >= 200 && response.status < 300) {
                        this.$emit('uploadSuccess', file.name);
                    } else {
                        this.$emit('uploadFailure', {
                            name: file.name,
                            message: `Received ${response.status} code from server.`,
                        });
                    }
                })
                .catch((error) => {
                    if (error.response) {
                        // The request was made and the server responded with a status code
                        // that falls out of the range of 2xx
                        this.$emit('uploadFailure', {
                            name: file.name,
                            message: `Received ${error.response.status} code from server.`,
                        });
                    } else if (error.request) {
                        // The request was made but no response was received
                        // `error.request` is an instance of XMLHttpRequest in the browser and an instance of
                        // http.ClientRequest in node.js
                        this.$emit('uploadFailure', { name: file.name, message: 'The server did not respond.' });
                    } else {
                        // Something happened in setting up the request that triggered an Error
                        this.$emit('uploadFailure', {
                            name: file.name,
                            message: 'There was an error sending your file to the server.',
                        });
                    }
                });
        },
    },
};
</script>

<style lang="scss" scoped>
@import '~@energysage/es-bs-base/scss/includes';

.es-file-upload {
    background-color: $gray-200;
    border: $btn-border-width dotted $border-color;
    @include media-breakpoint-down(sm) {
        background-color: $white;
        border: $btn-border-width solid $border-color;
    }

    &.active {
        border: $btn-border-width dotted $cyan-500;
    }
}

</style>
