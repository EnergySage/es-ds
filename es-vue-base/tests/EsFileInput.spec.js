import { mount } from '@vue/test-utils';
import EsFileInput from '@/src/lib-components/EsFileInput.vue';
import jestVue from '@/tests/jest.vue.config';

describe('EsFileInput', () => {
    // Basic test; should exist for most components
    test('<EsFileInput fileTypes={[\'image/png\']} />', async () => {
        const wrapper = mount(EsFileInput, {
            ...jestVue,
            propsData: {
                fileTypes: ['image/png'],
            },
        });

        // We do not check a11y here because the design of the component has nested interactive elements

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.props('fileTypes')).toStrictEqual(['image/png']);
        expect(wrapper.html()).toMatchSnapshot();
    });

    // Test slot is there
    test('<EsFileInput fileTypes={[\'image/png\']}>My Content</EsFileInput>', async () => {
        const wrapper = mount(EsFileInput, {
            ...jestVue,
            slots: {
                default: 'My Content',
            },
            propsData: {
                fileTypes: ['image/png'],
            },
        });

        expect(wrapper.vm.$slots.default[0].text).toBe('My Content');
        expect(wrapper.html()).toMatchSnapshot();
    });
});
