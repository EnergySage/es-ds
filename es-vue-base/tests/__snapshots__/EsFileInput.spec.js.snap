// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsFileInput <EsFileInput fileTypes={['image/png']} /> 1`] = `
"<div role="presentation" class="es-file-upload align-items-center rounded d-flex flex-column justify-content-center p-2">
  <div class="d-flex justify-content-center flex-column">
    <div class="align-self-center p-2"><svg viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg" style="height: 48px; width: 48px;" class="d-none d-md-inline-block mb-4">
        <path fill="currentColor" d="M1.5 30.734c.398 0 .78.182 1.06.506.282.324.44.763.44 1.22v8.634c0 .916.316 1.794.879 2.441.562.648 1.325 1.012 2.121 1.012h36c.796 0 1.559-.364 2.121-1.012.563-.647.879-1.525.879-2.441V32.46c0-.458.158-.897.44-1.221a1.408 1.408 0 0 1 1.06-.506c.398 0 .78.182 1.06.506.282.324.44.763.44 1.22v8.634c0 1.831-.632 3.588-1.757 4.883C45.117 47.272 43.59 48 42 48H6c-1.591 0-3.117-.728-4.243-2.023C.632 44.682 0 42.925 0 41.094V32.46c0-.458.158-.897.44-1.221a1.408 1.408 0 0 1 1.06-.506Z"></path>
        <path fill="currentColor" d="M22.94.507c.14-.16.305-.288.487-.375a1.331 1.331 0 0 1 1.15 0c.182.087.348.214.487.375l9 10.36c.282.324.44.764.44 1.222 0 .459-.158.898-.44 1.223a1.41 1.41 0 0 1-1.062.506 1.41 1.41 0 0 1-1.062-.506l-6.438-7.414V36.26c0 .458-.158.897-.44 1.221a1.408 1.408 0 0 1-1.06.506c-.398 0-.78-.182-1.06-.506a1.867 1.867 0 0 1-.44-1.22V5.897l-6.438 7.414c-.14.16-.305.287-.487.374a1.332 1.332 0 0 1-1.15 0 1.507 1.507 0 0 1-.487-.374c-.14-.161-.25-.352-.326-.561a1.958 1.958 0 0 1 0-1.323c.076-.21.186-.4.326-.561l9-10.36Z"></path>
      </svg></div>
    <div class="align-self-center p-2 text-center"></div>
    <div class="align-self-center p-2"><button type="button" class="btn d-none d-md-inline-block w-35 mb-4 btn-outline-secondary btn-md">
        Browse Files
      </button> <button type="button" class="btn d-md-none w-100 mb-2 btn-outline-secondary btn-md">
        Choose File
      </button></div>
  </div>
  <div></div>
  <div class="custom-file b-form-file is-valid" style="display: none;" id="__BVID__7__BV_file_outer_"><input type="file" accept="image/png" multiple="multiple" class="custom-file-input is-valid" style="z-index: -5;" id="__BVID__7"><label data-browse="Browse" class="custom-file-label" for="__BVID__7"><span class="d-block form-file-text" style="pointer-events: none;">No file chosen</span></label></div>
</div>"
`;

exports[`EsFileInput <EsFileInput fileTypes={['image/png']}>My Content</EsFileInput> 1`] = `
"<div role="presentation" class="es-file-upload align-items-center rounded d-flex flex-column justify-content-center p-2">
  <div class="d-flex justify-content-center flex-column">
    <div class="align-self-center p-2"><svg viewBox="0 0 48 48" fill="none" xmlns="http://www.w3.org/2000/svg" style="height: 48px; width: 48px;" class="d-none d-md-inline-block mb-4">
        <path fill="currentColor" d="M1.5 30.734c.398 0 .78.182 1.06.506.282.324.44.763.44 1.22v8.634c0 .916.316 1.794.879 2.441.562.648 1.325 1.012 2.121 1.012h36c.796 0 1.559-.364 2.121-1.012.563-.647.879-1.525.879-2.441V32.46c0-.458.158-.897.44-1.221a1.408 1.408 0 0 1 1.06-.506c.398 0 .78.182 1.06.506.282.324.44.763.44 1.22v8.634c0 1.831-.632 3.588-1.757 4.883C45.117 47.272 43.59 48 42 48H6c-1.591 0-3.117-.728-4.243-2.023C.632 44.682 0 42.925 0 41.094V32.46c0-.458.158-.897.44-1.221a1.408 1.408 0 0 1 1.06-.506Z"></path>
        <path fill="currentColor" d="M22.94.507c.14-.16.305-.288.487-.375a1.331 1.331 0 0 1 1.15 0c.182.087.348.214.487.375l9 10.36c.282.324.44.764.44 1.222 0 .459-.158.898-.44 1.223a1.41 1.41 0 0 1-1.062.506 1.41 1.41 0 0 1-1.062-.506l-6.438-7.414V36.26c0 .458-.158.897-.44 1.221a1.408 1.408 0 0 1-1.06.506c-.398 0-.78-.182-1.06-.506a1.867 1.867 0 0 1-.44-1.22V5.897l-6.438 7.414c-.14.16-.305.287-.487.374a1.332 1.332 0 0 1-1.15 0 1.507 1.507 0 0 1-.487-.374c-.14-.161-.25-.352-.326-.561a1.958 1.958 0 0 1 0-1.323c.076-.21.186-.4.326-.561l9-10.36Z"></path>
      </svg></div>
    <div class="align-self-center p-2 text-center"></div>
    <div class="align-self-center p-2"><button type="button" class="btn d-none d-md-inline-block w-35 mb-4 btn-outline-secondary btn-md">
        Browse Files
      </button> <button type="button" class="btn d-md-none w-100 mb-2 btn-outline-secondary btn-md">
        Choose File
      </button></div>
  </div>
  <div></div>
  <div class="custom-file b-form-file is-valid" style="display: none;"><input type="file" accept="image/png" multiple="multiple" class="custom-file-input is-valid" style="z-index: -5;"><label data-browse="Browse" class="custom-file-label"><span class="d-block form-file-text" style="pointer-events: none;">No file chosen</span></label></div>
</div>"
`;
